proposals/memory-specializations/utilities.tex

```diff
@@ -0,0 +1,349 @@
+\begin{codeblock}
+        for (; n > 0; (void)++first, --n)
+          ::new (static_cast<void*>(addressof(*first)))
+            @\changed{typename iterator_traits<ForwardIterator>::value_type}{value_type_t<O>}@;
+        return first;
+\end{codeblock}
+
+\rSec3[uninitialized.construct.value]{uninitialized_value_construct}
+\begin{codeblock}
+  template <@\changed{class ForwardIterator}{OutputIterator O, Sentinel<O> S}@>
+    @\added{requires Constructible<value_type_t<O>{>}() \&\&}@
+             @\added{Same<value_type_t<O>\&, reference_t<O>{>}}@
+    @\changed{void}{O}@ uninitialized_value_construct(@\changed{ForwardIterator}{O}@ first, @\changed{ForwardIterator}{S}@ last);
+\end{codeblock}
+
+
+\setcounter{Paras}{0}
+\effects Equivalent to:
+\begin{codeblock}
+        for (; first != last; ++first)
+          ::new (static_cast<void*>(addressof(*first)))
+            @\changed{typename iterator_traits<ForwardIterator>::value_type}{value_type_t<O>()}@;
+        @\added{return first;}@
+\end{codeblock}
+{\color{addclr}
+\begin{codeblock}
+  template <OutputRange Rng>
+    requires Constructible<value_type_t<iterator_t<Rng>>> &&
+             Same<value_type_t<iterator_t<O>>, reference_t<iterator_t<O>>
+  O uninitialized_value_construct(Rng&& rng);
+\end{codeblock}
+
+\pnum
+\effects Equivalent to:
+\begin{codeblock}
+      return uninitialized_value_construct(begin(rng), end(rng));
+\end{codeblock}
+} %% \color{addclr}
+
+\begin{codeblock}
+  template <@\changed{class}{OutputIterator}@ @\changed{OutputIterator}{O}\removed{, class Size}@>
+    @\added{requires Constructible<value_type_t<O>{>}() \&\&}@
+             @\added{Same<value_type_t<O>\&, reference_t<O>{>}}@
+  @\changed{ForwardIterator}{O}@
+    uninitialized_value_construct_n(@\changed{ForwardIterator}{O}@ first, @\changed{Size}{difference_type_t<O>}@ n);
+\end{codeblock}
+
+\pnum
+\effects Equivalent to:
+\begin{codeblock}
+        for (; n > 0; --n, (void)++first)
+          ::new (static_cast<void*>(addressof(*first)))
+            @\changed{typename iterator_traits<ForwardIterator>::value_type}{value_type_t<O>()}@;
+        @\added{return first;}@
+\end{codeblock}
+
+\rSec3[uninitialized.copy]{uninitialized_copy}
+\begin{codeblock}
+  template <@\changed{class}{InputIterator}@ @\changed{InputIterator}{I}@, @\added{Sentinel<I> S, }@
+            @\changed{class ForwardIterator}{ForwardIterator O}@>
+    @\added{requires }\added{Constructible<value_type_t<O>, reference_t<O>{>}() \&\&}@
+             @\added{Same<value_type_t<O>\&, reference_t<O>{>} \&\&}@
+             @\newtxt{CopyConstructible<value_type_t<O>{>}()}@
+  @\changed{ForwardIterator}{tagged_pair<tag::in(I), tag::out(O)>}@
+    uninitialized_copy(@\changed{InputIterator}{I}@ first, @\changed{InputIterator}{S}@ last, @\changed{ForwardIterator}{O}@ result);
+\end{codeblock}
+
+\setcounter{Paras}{0}
+\pnum
+\effects \changed{As if by}{Equivalent to}:
+\begin{codeblock}
+        for (; first != last; ++first)
```

Comment:
> Why do we need to cast `++first` to `void`?

1. Because this is quoting N4606 text and that's what N4606 says,
2. Since it's impossible for a function to have an argument of type `void`, converting an operand of the comma operator to `void` ensures that the expression will not invoke a user-defined overload of the comma operator.

>Why isn't it a named cast?

I assume simply for brevity.